rustls/src/msgs/codec.rs

```diff
@@ -174,88 +177,149 @@ impl Codec for u64 {
+        let mut sub = r.sub(len)?;
+        let mut items = Vec::with_capacity(len);
+        while sub.any_left() {
+            items.push(T::read(&mut sub)?);
+        }
 
-    for i in items {
-        i.encode(bytes);
+        Some(Self {
+            items,
+            len: PhantomData,
+        })
     }
+}
 
-    let len = bytes.len() - len_offset - 2;
-    debug_assert!(len <= 0xffff);
-    let out: &mut [u8; 2] = (&mut bytes[len_offset..len_offset + 2])
-        .try_into()
-        .unwrap();
-    *out = u16::to_be_bytes(len as u16);
+impl<L, A> FromIterator<A> for TlsVec<L, A> {
+    fn from_iter<T: IntoIterator<Item = A>>(iter: T) -> Self {
+        Vec::from_iter(iter).into()
+    }
 }
 
-pub fn encode_vec_u24<T: Codec>(bytes: &mut Vec<u8>, items: &[T]) {
-    let len_offset = bytes.len();
-    bytes.extend(&[0, 0, 0]);
+impl<'a, L, T> IntoIterator for &'a TlsVec<L, T> {
+    type Item = &'a T;
 
-    for i in items {
-        i.encode(bytes);
+    type IntoIter = std::slice::Iter<'a, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        (&self.items).iter()
     }
+}
+
+impl<L, T> Deref for TlsVec<L, T> {
+    type Target = Vec<T>;
 
-    let len = bytes.len() - len_offset - 3;
-    debug_assert!(len <= 0xff_ffff);
-    let len_bytes = u32::to_be_bytes(len as u32);
-    let out: &mut [u8; 3] = (&mut bytes[len_offset..len_offset + 3])
-        .try_into()
-        .unwrap();
-    out.copy_from_slice(&len_bytes[1..]);
+    fn deref(&self) -> &Self::Target {
+        &self.items
+    }
 }
 
-pub fn read_vec_u8<T: Codec>(r: &mut Reader) -> Option<Vec<T>> {
-    let mut ret: Vec<T> = Vec::new();
-    let len = usize::from(u8::read(r)?);
-    let mut sub = r.sub(len)?;
+impl<L, T> DerefMut for TlsVec<L, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.items
+    }
+}
 
-    while sub.any_left() {
-        ret.push(T::read(&mut sub)?);
+impl<L, T> From<TlsVec<L, T>> for Vec<T> {
+    fn from(tls: TlsVec<L, T>) -> Self {
+        tls.items
+    }
+}
+
+impl<L, T> From<Vec<T>> for TlsVec<L, T> {
+    fn from(items: Vec<T>) -> Self {
+        Self {
+            items,
+            len: PhantomData,
+        }
     }
+}
 
-    Some(ret)
+impl<L, T> Default for TlsVec<L, T> {
+    fn default() -> Self {
+        Self {
+            items: Default::default(),
+            len: Default::default(),
+        }
+    }
 }
 
-pub fn read_vec_u16<T: Codec>(r: &mut Reader) -> Option<Vec<T>> {
-    let mut ret: Vec<T> = Vec::new();
-    let len = usize::from(u16::read(r)?);
-    let mut sub = r.sub(len)?;
+const MAX_VEC_BYTES: usize = 0x1_0000;
 
-    while sub.any_left() {
-        ret.push(T::read(&mut sub)?);
+#[allow(clippy::use_self)]
+impl LengthPrefix for u8 {
+    fn write_into(len: usize, offset: usize, bytes: &mut [u8]) {
+        debug_assert!(len <= 0xff);
+        bytes[offset] = len as u8;
     }
 
-    Some(ret)
+    const BYTES: usize = 1;
+    const ZERO_LEN: &'static [u8] = &[0];
```

Comment:
Given `ZERO_LEN`, `BYTES` is just `ZERO_LEN.len()`. Conversely, given `BYTES`, `ZERO_LEN` is just `[0u8; Self::BYTES]`.

Either way, `write_into` can be implemented once given either `BYTES` or `ZERO_LEN`. So, I think the `LengthPrefix` trait should just have a single member: `const BYTES: usize`.